components/hab-spider/src/rdeps.rs

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) 2017 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{HashMap, HashSet};
+
+use petgraph::Graph;
+use petgraph::graph::NodeIndex;
+use petgraph::algo::{toposort, is_cyclic_directed};
+use petgraph::visit::{Bfs, Walker};
+
+#[derive(Debug, PartialEq)]
+pub enum GraphErr {
+    GraphCyclic,
+}
+
+pub type GType = usize;
+
+pub fn rdeps(g: &Graph<GType, GType>, n: NodeIndex) -> Result<Vec<GType>, GraphErr> {
+    if is_cyclic_directed(&g) {
+        error!("Input graph should not be cyclic!");
+        return Err(GraphErr::GraphCyclic);
+    }
+
+    // unwrap should never panic as we pre-check for cycle
+    let t: Vec<GType> = toposort(&g, None).unwrap().iter().map(|k| k.index()).collect();
+
+    let bfs: Vec<GType> = Bfs::new(&g, n).iter(&g).map(|k| k.index()).collect();
+
+    let mut topo_map: HashMap<usize, usize> = HashMap::new(); // Node -> Topo index
+
+    for (i, j) in t.iter().enumerate() {
+        topo_map.insert(*j, i);
+    }
+
+    // Collect BFS nodes in topological order
+    let start: usize = n.index();
+    let mut curr: usize = *topo_map.get(&start).unwrap(); // Where to start in topo array
```

Comment:
Yeah this is 'safe' in the sense that if the code gets to this point, it is guaranteed to not panic.  The only scenario is if the node index itself is 'bad', but if that is the case Bfs::new will panic.  Since node indices are generated only by the graph, we'd have to make an effort to try and create bad indices to make that happen.